File: api/schemas/envelopes.py
"""Pydantic schemas for envelope budgeting ("Sobres").

Spending-cap envelopes: user-named buckets tied to a class
(needs/wants/savings/investing) with a monthly limit. Spend is computed live
from transactions; the summary is the home-tab feed. Amounts are floats in the
API (matching debts/transactions), stored as NUMERIC.
"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field

EnvelopeClass = Literal["needs", "wants", "savings", "investing"]
CurrencyLit = Literal["CRC", "USD"]


class EnvelopeCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=120)
    envelope_class: EnvelopeClass
    limit_amount: float = Field(..., gt=0)
    currency: CurrencyLit = "CRC"
    # Phase 7a: when set, this envelope nests under `parent_id`. The child
    # inherits the parent's class + currency (the values above are ignored for
    # children, validated server-side), and depth = parent.depth + 1 (cap 5).
    parent_id: Optional[uuid.UUID] = None


class EnvelopeUpdate(BaseModel):
    """Narrowed whitelist. `currency` and `period` are immutable post-create
    (changing currency would make a stored limit ambiguous against historical
    spend). `parent_id` is NOT here — re-parenting is out of scope (it would
    force a subtree depth recompute). Editing `envelope_class` is allowed only
    on a root and cascades to the subtree (enforced in the router)."""

    model_config = {"extra": "forbid"}

    name: Optional[str] = Field(None, min_length=1, max_length=120)
    envelope_class: Optional[EnvelopeClass] = None
    limit_amount: Optional[float] = Field(None, gt=0)
    is_active: Optional[bool] = None
    archived: Optional[bool] = None


class EnvelopeResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    parent_id: Optional[uuid.UUID] = None
    depth: int
    name: str
    envelope_class: str
    limit_amount: float
    currency: str
    period: str
    is_active: bool
    archived: bool
    created_at: datetime
    updated_at: datetime
    # Shared envelopes (defaulted; set only when this row is returned to a
    # MEMBER via GET /envelopes). `is_shared=True` means the caller is a member,
    # not the owner — `user_id` is the owner's id. Own rows leave these defaulted.
    is_shared: bool = False
    role: Optional[str] = None  # "member" when shared
    shared_by_name: Optional[str] = None
    member_count: int = 0

    model_config = {"from_attributes": True}


# ── summary (home-tab feed) ───────────────────────────────────────────────────


class EnvelopeSummaryItem(BaseModel):
    id: uuid.UUID
    parent_id: Optional[uuid.UUID] = None
    depth: int = 1
    name: str
    envelope_class: str
    currency: str
    limit_amount: float
    # `spent` is the ROLLED-UP figure (this node + all descendants) — what the
    # bar shows. `direct_spent` is only what's tagged directly to this node.
    spent: float
    direct_spent: float
    # B2 fixed-expense attachment: `reserved` = Σ attached bills/debts unpaid this
    # cycle (rolled up); `available` = limit − reserved − spent (the truly free
    # amount). `remaining` (limit − spent) stays for back-compat.
    reserved: float = 0.0
    available: float = 0.0
    remaining: float  # limit − rolled-up spent
    pct: float  # rolled-up spent / limit, 0..>1 (clamp in UI; >1 = over)
    over_limit: bool
    # Soft allocation (Phase 7a): how the node's budget is split across its
    # direct children. `unallocated` may be negative (over-allocated).
    allocated: float = 0.0
    unallocated: float = 0.0
    over_allocated: bool = False
    # Shared envelopes: when `is_shared`, this item belongs to ANOTHER user's
    # tree the caller is a member of. `spent` is then the SHARED (all-members)
    # rolled-up figure and `your_spent` is the caller's own portion ("Vos: ₡X de
    # ₡Y"). Shared items are display-only — they are NOT counted in by_class /
    # total_* and the byte-locked cashflow/affordability/snapshot consumers
    # filter them out. Defaulted so own-envelope serialization is unchanged.
    is_shared: bool = False
    role: Optional[str] = None  # "member" on shared items; None on own
    shared_by_name: Optional[str] = None
    member_count: int = 0
    your_spent: float = 0.0


class EnvelopeClassSubtotal(BaseModel):
    envelope_class: str
    limit_total: float
    spent_total: float
    over_limit: bool


class EnvelopeSummaryResponse(BaseModel):
    period: str  # e.g. "2026-06"
    currency: str
    envelopes: list[EnvelopeSummaryItem] = Field(default_factory=list)
    by_class: list[EnvelopeClassSubtotal] = Field(default_factory=list)
    total_limit: float
    # Grand totals across all envelopes in the summary currency (roots only,
    # like total_limit, so subtrees aren't double-counted). They power the
    # native home-screen "Te queda este mes" headline; total_available may be
    # negative when the overall budget is blown.
    total_spent: float = 0.0
    total_reserved: float = 0.0
    total_available: float = 0.0
    monthly_income: Optional[float] = None  # for the "split income" sanity line

File: mobile/src/api/envelopes.ts
import { api } from "./client";

// Envelope budgeting ("Sobres") — spending-cap buckets. Spend is computed
// live on the backend (no stored balance); this client just renders it.

export type EnvelopeClass = "needs" | "wants" | "savings" | "investing";

export interface EnvelopeResponse {
  id: string;
  user_id: string;
  parent_id: string | null;
  depth: number;
  name: string;
  envelope_class: EnvelopeClass;
  limit_amount: number;
  currency: string;
  period: string;
  is_active: boolean;
  archived: boolean;
  created_at: string;
  updated_at: string;
}

export interface EnvelopeCreate {
  name: string;
  envelope_class: EnvelopeClass;
  limit_amount: number;
  currency?: "CRC" | "USD";
  // Phase 7a: when set, nests this envelope under parent_id. The child inherits
  // the parent's class + currency server-side (the values above are ignored).
  parent_id?: string | null;
}

/**
 * Minimal shape the edit sheet needs. Both a full `EnvelopeResponse` and an
 * `EnvelopeSummaryItem` satisfy it, so the home-tab section can open the editor
 * straight from the summary it already has — no second fetch, no dead taps.
 */
export type EnvelopeEditable = Pick<
  EnvelopeResponse,
  "id" | "name" | "envelope_class" | "limit_amount" | "currency"
>;

export interface EnvelopeUpdate {
  name?: string;
  envelope_class?: EnvelopeClass;
  limit_amount?: number;
  is_active?: boolean;
  archived?: boolean;
}

// ── summary (home-tab feed) ─────────────────────────────────────────────────

export interface EnvelopeSummaryItem {
  id: string;
  parent_id: string | null;
  depth: number;
  name: string;
  envelope_class: EnvelopeClass;
  currency: string;
  limit_amount: number;
  spent: number; // rolled-up (this node + descendants) — the bar value
  direct_spent: number; // only what's tagged directly to this node
  // B2 fixed-expense attachment: `reserved` = attached bills/debts unpaid this
  // cycle; `available` = limit − reserved − spent (the truly free amount).
  reserved: number;
  available: number;
  remaining: number; // limit − spent (back-compat)
  pct: number; // rolled-up spent / limit; clamp in UI; > 1 = over
  over_limit: boolean;
  // Soft allocation: how this node's budget is split across direct children.
  allocated: number;
  unallocated: number; // limit − allocated; may be negative (over-allocated)
  over_allocated: boolean;
}

export interface EnvelopeClassSubtotal {
  envelope_class: EnvelopeClass;
  limit_total: number;
  spent_total: number;
  over_limit: boolean;
}

export interface EnvelopeSummaryResponse {
  period: string;
  currency: string;
  envelopes: EnvelopeSummaryItem[];
  by_class: EnvelopeClassSubtotal[];
  total_limit: number;
  // Phase 7c: grand totals (roots only, summary currency) — the home-screen
  // "Te queda este mes" headline. total_available can be negative.
  total_spent: number;
  total_reserved: number;
  total_available: number;
  monthly_income: number | null;
}

// ── display metadata (CR Spanish, neutral palette per theme.ts v2) ──────────

export const ENVELOPE_CLASS_ORDER: EnvelopeClass[] = [
  "needs",
  "wants",
  "savings",
  "investing",
];

export const ENVELOPE_CLASS_LABELS: Record<EnvelopeClass, string> = {
  needs: "Necesidades",
  wants: "Gustos",
  savings: "Ahorro",
  investing: "Inversión",
};

export const ENVELOPE_CLASS_COLORS: Record<EnvelopeClass, string> = {
  needs: "#3E7A57", // green
  wants: "#A8763B", // amber
  savings: "#33708A", // steel teal
  investing: "#6E5A93", // muted violet
};

// The bar shows money LEFT: it starts full (100%) and drains with each expense.
// It goes red in the last 5% (and stays red once over the limit).
export const ENVELOPE_LOW_THRESHOLD = 0.05;

export interface EnvelopeProgress {
  available: number; // limit − reserved − spent (the truly free money) — bar fill
  remaining: number; // back-compat alias for `available` (existing callers)
  reserved: number; // attached fixed expenses, unpaid this cycle
  spent: number; // actual spend
  fraction: number; // available / limit as 0..1 (clamped) — the green segment
  reservedFraction: number; // reserved / limit as 0..1 — the reserved segment
  spentFraction: number; // spent / limit as 0..1 — the spent segment
  low: boolean; // available ≤ 5% of limit → red
}

export function envelopeProgress(
  item: Pick<EnvelopeSummaryItem, "limit_amount" | "remaining"> &
    Partial<Pick<EnvelopeSummaryItem, "spent" | "reserved" | "available">>
): EnvelopeProgress {
  const limit = item.limit_amount;
  const reserved = item.reserved ?? 0;
  const spent = item.spent ?? 0;
  // Back-compat: callers that only pass `remaining` (limit − spent) get that as
  // `available` (no reserved segment) until they pass the full summary item.
  const available = item.available ?? item.remaining;
  const clamp = (v: number) => (limit > 0 ? Math.max(0, Math.min(v / limit, 1)) : 0);
  return {
    available,
    remaining: available,
    reserved,
    spent,
    fraction: clamp(available),
    reservedFraction: clamp(reserved),
    spentFraction: clamp(spent),
    low: limit > 0 ? available <= ENVELOPE_LOW_THRESHOLD * limit : available <= 0,
  };
}

// ── nesting (Phase 7a) ───────────────────────────────────────────────────────
// The summary + list endpoints return a FLAT list linked by parent_id. Build
// the tree client-side, then flatten in pre-order so a parent renders right
// above its sub-sobres (indent by `depth`).

interface Nestable {
  id: string;
  parent_id: string | null;
}

export function flattenEnvelopeTree<T extends Nestable>(items: T[]): T[] {
  const childrenOf = new Map<string | null, T[]>();
  const ids = new Set(items.map((it) => it.id));
  for (const it of items) {
    // Treat a node whose parent isn't in this set (e.g. archived) as a root.
    const key = it.parent_id && ids.has(it.parent_id) ? it.parent_id : null;
    const list = childrenOf.get(key) ?? [];
    list.push(it);
    childrenOf.set(key, list);
  }
  const out: T[] = [];
  const walk = (parentId: string | null) => {
    for (const it of childrenOf.get(parentId) ?? []) {
      out.push(it);
      walk(it.id);
    }
  };
  walk(null);
  return out;
}

// ── API ─────────────────────────────────────────────────────────────────────

export async function fetchEnvelopes(
  includeArchived = false
): Promise<EnvelopeResponse[]> {
  const res = await api.get<EnvelopeResponse[]>("/envelopes", {
    params: includeArchived ? { include_archived: true } : undefined,
  });
  return res.data;
}

export async function fetchEnvelopeSummary(): Promise<EnvelopeSummaryResponse> {
  const res = await api.get<EnvelopeSummaryResponse>("/envelopes/summary");
  return res.data;
}

export async function createEnvelope(
  payload: EnvelopeCreate
): Promise<EnvelopeResponse> {
  const res = await api.post<EnvelopeResponse>("/envelopes", payload);
  return res.data;
}

export async function updateEnvelope(
  id: string,
  payload: EnvelopeUpdate
): Promise<EnvelopeResponse> {
  const res = await api.patch<EnvelopeResponse>(`/envelopes/${id}`, payload);
  return res.data;
}

/** Soft archive — hides the envelope and stops it counting; tagged
 * transactions keep their link. */
export async function archiveEnvelope(id: string): Promise<EnvelopeResponse> {
  const res = await api.delete<EnvelopeResponse>(`/envelopes/${id}`);
  return res.data;
}

/** Hard delete — permanently removes the envelope. Tagged transactions are
 * unlinked (envelope_id → NULL), never deleted. */
export async function deleteEnvelope(id: string): Promise<EnvelopeResponse> {
  const res = await api.delete<EnvelopeResponse>(`/envelopes/${id}`, {
    params: { hard: true },
  });
  return res.data;
}

/**
 * Assign (or clear, with `null`) the envelope an expense counts against.
 * Goes through PATCH /transactions/{id} — the backend validates the envelope
 * belongs to the caller and is not archived (400 "Sobre inválido." otherwise),
 * and 409s on shadow/transfer/archived rows.
 */
export async function assignTransactionEnvelope(
  transactionId: string,
  envelopeId: string | null
): Promise<void> {
  await api.patch(`/transactions/${transactionId}`, {
    envelope_id: envelopeId,
  });
}
